Add agent framework/throttling/hidden model/OS assistant and update conversational search documentation (#6354)
opensearch-project/documentation-website: _ml-commons-plugin/api/memory-apis/get-memory.md

---
layout: default
title: Get memory
parent: Memory APIs
grand_parent: ML Commons APIs
nav_order: 20
---

# Get a memory
**Introduced 2.12**
{: .label .label-purple }

> Retrieves a conversational memory for [conversational search]({{site.url}}{{site.baseurl}}/search-plugins/conversational-search/). 

Author: ```suggestion
Use this API to retrieve a conversational memory for [conversational search]({{site.url}}{{site.baseurl}}/search-plugins/conversational-search/). 
```

### Paginating results

The `next_token` parameter provides the ordered position of the first memory within the sorted list of memories to return in the results. When a memory is added between subsequent GET Memory calls, one of the listed memories will be duplicated in the results. For example, suppose the current ordered list of memories is `BCDEF`, where `B` is the memory created most recently. When you call the Get Memory API with `next_token=0` and `max_results=3`, the API returns `BCD`. Let's say you then create another memory A. The memory list now looks like `ABCDEF`. The next time you call the Get Memory API with `next_token=3` and `max_results=3`, you'll receive `DEF` in the results. Notice that `D` will be returned in the first and second batches of results. The following diagram illustrates the duplication.

Request | List of memories (returned memories are enclosed in brackets) | Results returned in the response
:--- | :--- | :---
Get Memory (next_token = 0, max_results = 3) | [BCD]EF | BCD
Create Memory            | ABCDEF | -

> Get Memory(next_token = 3, max_results = 3) -> ABC[DEF] | DEF

Author: ```suggestion
Get Memory (next_token = 3, max_results = 3) -> ABC[DEF] | DEF
```

#### Example request: Get all memories

```json
GET /_plugins/_ml/memory/
```
{% include copy-curl.html %}

#### Example request: Paginating results

```json
GET /_plugins/_ml/memory?max_results=2&next_token=1
```

#### Example response

```json
{
  "memories": [
    {
      "memory_id": "gW8Aa40BfUsSoeNTvOKI",
      "create_time": "2024-02-02T18:07:06.887061463Z",
      "updated_time": "2024-02-02T19:01:32.121444968Z",
      "name": "Conversation for a RAG pipeline",
      "user": "admin"
    }
  ]
}
```

## Response fields

> The following table lists all response fields.

Author: ```suggestion
The following table lists the available response fields.
```

```
### Path parameters

The following table lists the available path parameters.

Parameter | Data type | Description
:--- | :--- | :---
`memory_id` | String | The ID of the memory to retrieve.

#### Example request

```json
GET /_plugins/_ml/memory/N8AE1osB0jLkkocYjz7D
```
{% include copy-curl.html %}

#### Example response

```json
{
  "memory_id": "gW8Aa40BfUsSoeNTvOKI",
  "create_time": "2024-02-02T18:07:06.887061463Z",
  "updated_time": "2024-02-02T19:01:32.121444968Z",
  "name": "Conversation for a RAG pipeline, updated",
  "user": "admin"
}
```

## Search for a memory

Use this command to search for memories.

### Path and HTTP methods

```json
GET /_plugins/_ml/memory/_search
POST /_plugins/_ml/memory/_search
```

#### Example request: Searching for all memories

```json
POST /_plugins/_ml/memory/_search
{
  "query": {
    "match_all": {}
  },
  "size": 1000
}
```
{% include copy-curl.html %}

#### Example request: Searching for memory by name

```json
POST /_plugins/_ml/memory/_search
{
  "query": {
    "term": {
      "name": {
        "value": "conversation"
      }
    }
  }
}
```
{% include copy-curl.html %}

#### Example response

```json
{

>   "conversations": [

Reviewer: This is the response of List_Memory, not the search memory. Also, the list memory should have "memories" rather than "conversations".

Author: This is because these two APIs are combined on one page. I will separate the search and get APIs for all applicable APIs.

Request | List of memories (returned memories are enclosed in brackets) | Results returned in the response
:--- | :--- | :---
Get Memory (next_token = 0, max_results = 3) | [BCD]EF | BCD
Create Memory            | ABCDEF | -
Get Memory(next_token = 3, max_results = 3) -> ABC[DEF] | DEF

#### Example request: Searching for all memories

```json
GET /_plugins/_ml/memory/
```
{% include copy-curl.html %}

#### Example request: Paginating results

```json
GET /_plugins/_ml/memory?max_results=2&next_token=1
```

#### Example response

```json
{

>   "conversations": [

Reviewer: Here the name should be "memories". Example:
{
    "memories": [
        {
            "memory_id": "XH94pI0Bc7x8SkEm2eCU",
            "create_time": "2024-02-13T21:56:39.950289Z",
            "updated_time": "2024-02-13T21:56:39.950289Z",
            "name": "Test conversation for RAG pipeline"
        }
    ]
}
